pywhy_stats/custom_kernels.py

```diff
@@ -0,0 +1,91 @@
+from typing import Union
+
+import numpy as np
+from numpy import exp, median, shape, sqrt, ndarray
+from numpy.random import permutation
+from scipy.spatial.distance import squareform, pdist
+
+
+def rbf_kernel(X: np.ndarray, width: Union[float, str] = "median") -> np.ndarray:
+    """
+    Custom implementation of the RBF kernel.
+
+    Parameters
+    ----------
+    X: numpy array of shape (n_samples, n_columns)
+    width : string or float
+            The width for the RBF kernel. This can either be provided as a float or as a string indicating one of
+            the following approximation methods:
+                'empirical_hsic': using empirical rule for HSIC
+                'empirical_kci': using empirical rule for KCI
+                'median': using the median trick
+
+    Returns
+    -------
+    result: numpy array of shape (n_samples, n_samples)
+            The resulting kernel matrix after applying the RBF kernel.
+    """
+    if isinstance(width, str):
+        if width == "median":
+            width = _get_gaussian_median_width(X)
+        elif width == "empirical_hsic":
+            width = _get_gaussian_empirical_hsic(X)
+        elif width == "empirical_kci":
+            width = _get_gaussian_empirical_kci(X)
+        else:
+            raise ValueError("Unknown width approximation method!")
+
+    return exp(-0.5 * squareform(pdist(X, 'sqeuclidean')) * width)
+
+
+def delta_kernel(X: np.ndarray) -> np.ndarray:
+    """
+    Delta kernel for categorical values. This is, the similarity is 1 if the values are equal and 0 otherwise.
+
+    Parameters
+    ----------
+    X: numpy array of shape (n_samples, n_columns)
+
+    Returns
+    -------
+    result: numpy array of shape (n_samples, n_samples)
+            The resulting kernel matrix after applying the delta kernel.
+    """
+    if X.ndim == 1:
+        X = X.reshape(-1, 1)
+
+    return np.array(list(map(lambda value: value == X, X))).reshape(X.shape[0], X.shape[0]).astype(np.float32)
+
+
+def _get_gaussian_median_width(X: np.ndarray) -> float:
+    n = shape(X)[0]
+    if n > 1000:
+        X = X[permutation(n)[:1000], :]
+    dists = squareform(pdist(X, 'euclidean'))
+    median_dist = median(dists[dists > 0])
+    width = sqrt(2.) * median_dist
+    return 1.0 / (width ** 2)
+
+
+def _get_gaussian_empirical_kci(X: np.ndarray) -> float:
+    n = shape(X)[0]
+    if n < 200:
+        width = 1.2
+    elif n < 1200:
+        width = 0.7
+    else:
+        width = 0.4
+    theta = 1.0 / (width ** 2)
+    return theta / X.shape[1]
```

Comment:
Do we have the exps to back up these numbers? If so, it would be great to also port this into a `benchmarks/` folder and add a link to relevant paper/webpage/etc.

If not, then I would be cautious cuz idk where these came from, so some user might ask us why it defaults to these values and we would have no idea.

Reply:
These are just copied over form the original implementation https://github.com/py-why/causal-learn/blob/main/causallearn/utils/KCI/GaussianKernel.py#L39. As far as I know, they were already used in the origin implementation from R.

Reply:
Tagging @kunwuz.

Any idea where these values came from? Anyway, we can inform ppl?

Reply:
They are from the second paragraph of section 3.5 in the [KCI paper ](https://arxiv.org/ftp/arxiv/papers/1202/1202.3775.pdf). I agree that including a reference could be helpful for these empirical values.